internal/flowcontrol/writebuffer.go

```diff
@@ -207,3 +208,44 @@ type jsonPatch struct {
 	Path  string `json:"path"`
 	Value any    `json:"value"`
 }
+
+type rateLimiter struct {
+	failuresLock sync.Mutex
+	failures     map[interface{}]int
+	limiter      *rate.Limiter
+}
+
+func newRateLimiter(batchInterval time.Duration, burst int) workqueue.RateLimiter {
+	return &rateLimiter{
+		failures: map[interface{}]int{},
+		limiter:  rate.NewLimiter(rate.Every(batchInterval), burst),
+	}
+}
+
+func (r *rateLimiter) When(item any) time.Duration {
+	r.failuresLock.Lock()
+	defer r.failuresLock.Unlock()
+
+	failures := r.failures[item]
+	r.failures[item]++
+
+	// Retry quickly a few times using exponential backoff
+	if failures > 0 && failures < 5 {
+		return time.Duration(10*math.Pow(2, float64(failures-1))) * time.Millisecond
+	}
+
+	// Non-error batching interval
+	return r.limiter.Reserve().Delay()
+}
+
+func (r *rateLimiter) NumRequeues(item interface{}) int {
+	r.failuresLock.Lock()
```

Comment:
RWMutex?

Reply:
NumRequeues shouldn't actually be called anywhere since nothing wraps this rate limiter impl. And forget is called from the same goroutine as When. So I don't think there's really a perf upside